code/game/objects/items/RTD.dm

```diff
@@ -0,0 +1,359 @@
+ * for basic directions its rotated with CSS so there is no need for icon
+ */
+/datum/tile_info/proc/get_icon_state()
+	var/prefix = ""
+	if(selected_direction)
+		prefix = (selected_direction in GLOB.tile_dont_rotate) ? "" : "-[dir2text(selected_direction)]"
+	return icon_state + prefix
+
+///convinience proc to quickly convert the tile design into an physical tile to lay on the plating
+/datum/tile_info/proc/new_tile(loc)
+	var/obj/item/stack/tile/final_tile = new tile_type(loc, 1)
+	final_tile.turf_dir = selected_direction
+	return final_tile
+
+/**
+ * Stores the decal & overlays on the floor to preserve texture of the design
+ * in short its just an wrapper for mutable appearance where we retrive the nessassary information
+ * to recreate an mutable appearance
+ */
+/datum/overlay_info
+	/// icon var of the mutable appearance
+	var/icon/icon
+	/// icon_state var of the mutable appearance
+	var/icon_state
+	/// direction var of the mutable appearance
+	var/direction
+	/// alpha var of the mutable appearance
+	var/alpha
+	/// color var of the mutable appearance
+	var/color
+
+//decompressing nessasary information required to re-create an mutable appearance
+/datum/overlay_info/New(mutable_appearance/appearance)
+	icon = appearance.icon
+	icon_state = appearance.icon_state
+	alpha = appearance.alpha
+	direction = appearance.dir
+	color = appearance.color
+
+/// re create the appearance
+/datum/overlay_info/proc/add_decal(turf/the_turf)
+	the_turf.AddElement(/datum/element/decal, icon, icon_state, direction, null, null, alpha, color, null, FALSE, null)
+
+/obj/item/construction/rtd/Initialize(mapload)
+	. = ..()
+	selected_design = new
+	tile_design = new
+
+	selected_design.set_info(GLOB.floor_designs[root_category][design_category][1])
+	update_appearance()
+
+/obj/item/construction/rtd/Destroy()
+	QDEL_NULL(selected_design)
+	QDEL_NULL(tile_design)
+	QDEL_LIST(design_overlays)
+	. = ..()
+
+/obj/item/construction/rtd/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "RapidTilingDevice", name)
+		ui.open()
+
+
+/obj/item/construction/rtd/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/rtd),
+	)
+
+/obj/item/construction/rtd/attack_self(mob/user)
+	. = ..()
+	ui_interact(user)
+
+/obj/item/construction/rtd/ui_data(mob/user)
+	var/list/data = ..()
+	var/floor_designs = GLOB.floor_designs
+
+	data["selected_root"] = root_category
+	data["root_categories"] = list()
+	for(var/category in floor_designs)
+		data["root_categories"] += category
+	data["selected_category"] = design_category
+
+	selected_design.fill_ui_data(data)
+
+	data["categories"] = list()
+	for(var/sub_category as anything in floor_designs[root_category])
+		var/list/target_category =  floor_designs[root_category][sub_category]
+
+		var/list/designs = list() //initialize all designs under this category
+		for(var/i in 1 to target_category.len)
```

Comment:
Can you use a `for var/category in target_category` loop here?

Reply:
Yup